### PR TITLE
Final integration

### PR DESCRIPTION
**Pull request by jhennawi**
This is a reasonably sized pull request that involves changes resulting from trying to make the master branch more compatible with upcoming implementation of revamped object finding and extraction. 

We also add the functionality to allow for wavetilts to take different line tracing thresholds for different orders/slits. This is implemented via optionally making one of the settings a list rather than a float. It will run correctly either way. 

Other fixes to extraction related to differences in PYPIT read noise image definition (rn2 rather than rn) have been implemented as these were bugs in the first version of the extraction algorithms.

### CODE REVIEW

**Comment by profxj** [comment]
Please begin by editing the CHANGES file, @jhennawi 

Thanks.

File: pypit/core/ararc.py [hunk 1]
```diff
@@ -172,6 +172,7 @@ def detect_lines(censpec, nfitpix=5, nonlinear=None):
       The spectrum used to find detections. This spectrum has
       had any "continuum" emission subtracted off
     """
+    # TODO I don't see that there is any continuum subtraction being done here contrary to what the docs say.
```
Comment (profxj) [inline pypit/core/ararc.py]: There might have been previously.  I probably removed it.

File: pypit/core/arextract.py [hunk 2]
```diff
@@ -239,15 +239,15 @@ def extract_boxcar(image,trace_in, radius_in, ycen = None):
     return fextract
 
 
-def extract_optimal(waveimg, imgminsky, ivar, mask, oprof, skyimg, rn_img, box_radius, specobj):
+def extract_optimal(waveimg, imgminsky, ivar, mask, oprof, skyimg, rn2_img, box_radius, specobj):
```
Comment (profxj) [inline pypit/core/arextract.py]: I think you need to update the doc string too.

File: pypit/core/arprocimg.py [hunk 3]
```diff
@@ -772,8 +773,8 @@ def variance_frame(datasec_img, det, sciframe, settings_det=None,
         if dnoise is None:
             dnoise = (settings_det['darkcurr'] * float(fitsdict["exptime"][idx])/3600.0)
         # Return
```
Comment (profxj) [inline pypit/core/arprocimg.py]: Add a doc string explaining this kludge as per @rcooke-ast 's request.

File: pypit/core/arskysub.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 
 def bg_subtraction_slit(slit, slitpix, edge_mask, sciframe, varframe, tilts,
```
Comment (profxj) [inline pypit/core/arskysub.py]: Add new params to doc string

File: pypit/core/arskysub.py [hunk 5]
```diff
@@ -80,33 +80,34 @@ def bg_subtraction_slit(slit, slitpix, edge_mask, sciframe, varframe, tilts,
...
-            debugger.set_trace()
-        lskyset, outmask, lsky_fit, red_chi = arutils.bspline_profile(
-            wsky[pos_sky], lsky, lsky_ivar, np.ones_like(lsky),
-            fullbkpt = tmp.breakpoints, upper=sigrej, lower=sigrej,
-            kwargs_reject={'groupbadpix':True})
-        res = (sky[pos_sky] - np.exp(lsky_fit)) * np.sqrt(sky_ivar[pos_sky])
-        lmask = (res < 5.0) & (res > -4.0)
-        sky_ivar[pos_sky] = sky_ivar[pos_sky] * lmask
+    # Restrict fit to positive pixels only and mask out large outliers via a pre-fit to the log
+    if (POS_MASK==True):
```
Comment (profxj) [inline pypit/core/arskysub.py]:
```
POS_MASK is True
```

File: pypit/core/arskysub.py [hunk 6]
```diff
@@ -118,13 +119,18 @@ def bg_subtraction_slit(slit, slitpix, edge_mask, sciframe, varframe, tilts,
     bgframe[all_slit] = skyset.value(piximg[all_slit])[0] #, skyset)
 
     # Debugging/checking
-    if False:
+    if PLOT_FIT:
         from matplotlib import pyplot as plt
+        from IPython import embed
```
Comment (profxj) [inline pypit/core/arskysub.py]: IPython is not included in our dependencies..  Remove

File: pypit/core/arskysub.py [hunk 7]
```diff
@@ -118,13 +119,18 @@ def bg_subtraction_slit(slit, slitpix, edge_mask, sciframe, varframe, tilts,
     bgframe[all_slit] = skyset.value(piximg[all_slit])[0] #, skyset)
 
     # Debugging/checking
-    if False:
+    if PLOT_FIT:
         from matplotlib import pyplot as plt
```
Comment (profxj) [inline pypit/core/arskysub.py]: All imports should be at the top of the module.

I might suggest you move this to ```ardebug```

File: pypit/core/artraceslits.py [hunk 8]
```diff
@@ -2802,51 +2807,97 @@ def trace_gweight(fimage, xinit_in, sigma = 1.0, ycen = None, invvar=None, maskv
...
 
-    # Setup
+
+    # Init
     nx = fimage.shape[1]
-    xnew = np.zeros_like(xcen)
-    xerr = maskval*np.ones_like(xnew)
+
+    # Checks on radius
+    if (isinstance(sigma,int) or isinstance(sigma,float)):
```
Comment (profxj) [inline pypit/core/artraceslits.py]: cleaner as:

```
isinstance(sigma, (int,float))
```

File: pypit/core/artraceslits.py [hunk 9]
```diff
@@ -2802,51 +2807,97 @@ def trace_gweight(fimage, xinit_in, sigma = 1.0, ycen = None, invvar=None, maskv
...
+        sigma_out = sigma
+    else:
+        raise ValueError('Gaussian sigma must a be either an integer, a floating point number, or an ndarray '
+                         'with the same shape and size as xinit_in')
+
+
+    # Figure out dimensions of xinit
+    dim = xinit_in.shape
+    ndim = xinit_in.ndim
+    if (ndim == 1):
```
Comment (profxj) [inline pypit/core/artraceslits.py]: Change the doc string to indicate the multiple ways xinit 
may be input.

File: pypit/core/artraceslits.py [hunk 10]
```diff
@@ -2802,51 +2807,97 @@ def trace_gweight(fimage, xinit_in, sigma = 1.0, ycen = None, invvar=None, maskv
...
+    ncen = xinit_in.size
+
+    xinit = xinit_in.flatten()
+    # Create xnew, xerr
+    xnew = xinit.astype(float)
+    xerr = np.full(ncen,999.)
+
+    if ycen is None:
+        if ndim == 1:
+            ycen = np.arange(npix, dtype='int')
```
Comment (profxj) [inline pypit/core/artraceslits.py]: I still don't understand why one puts '' around int.
I'd avoid it.

File: pypit/core/artraceslits.py [hunk 11]
```diff
@@ -2802,51 +2807,97 @@ def trace_gweight(fimage, xinit_in, sigma = 1.0, ycen = None, invvar=None, maskv
...
+        elif ndim == 2:
+            ycen = np.outer(np.arange(npix, dtype='int'), np.ones(nTrace, dtype='int'))
+        else:
+            raise ValueError('xinit is not 1 or 2 dimensional')
```
Comment (profxj) [inline pypit/core/artraceslits.py]: Replace all ```raise ValueError``` in this PR by
```msgs.error```

File: pypit/core/artracewave.py [hunk 12]
```diff
@@ -171,6 +171,7 @@ def trace_tilt(ordcen, rordloc, lordloc, det, msarc, slitnum, settings_det,
     This function performs a PCA analysis on the arc tilts for a single spectrum (or order)
                tracethresh=1000.0, nsmth=0):
 
+    # TODO These docs are pathetic, please elaborate.
```
Comment (profxj) [inline pypit/core/artracewave.py]: I find the use of 'pathetic' to be offensive.  Remove.

File: pypit/ginga.py [hunk 13]
```diff
@@ -124,7 +124,7 @@ def show_slits(viewer, ch, lord_in, rord_in, slit_ids = None, rotate=False, pste
         lordloc = lord_in.reshape(lord_in.size,1)
         rordloc = rord_in.reshape(rord_in.size,1)
 
-    if slit_ids == None:
+    if np.any(slit_ids) == None:
```
Comment (profxj) [inline pypit/ginga.py]: Will ```np.any()``` ever return ```None```?  I doubt it.
This needs fixing.

File: pypit/scienceimage.py [hunk 14]
```diff
@@ -455,15 +456,19 @@ def global_skysub(self, settings_skysub, use_tracemask=False):
...
         for slit in gdslits:
             msgs.info("Working on slit: {:d}".format(slit))
             # Find sky
             slit_bgframe = arskysub.bg_subtraction_slit(
                 slit+1, self.tslits_dict['slitpix'], self.tslits_dict['edge_mask'],
                 self.sciframe, varframe, self.tilts,
                 bpm=self.bpm, crmask=self.crmask, tracemask=tracemask)
-            # Add
-            self.global_sky += slit_bgframe
+            # Mask?
```
Comment (profxj) [inline pypit/scienceimage.py]: Are we ok with this Hack I put in?

File: pypit/traceslits.py [hunk 15]
```diff
@@ -155,6 +155,7 @@ def __init__(self, mstrace, pixlocn, binbpx=None, settings=None, det=None,
         self.lcen = None     # narray
         self.rcen = None     # narray
         self.tc_dict = None  # dict
+        # TODO Why is there a tc_dict and a tslits_dict. This is confusing naming without some docs or comments here.
```
Comment (profxj) [inline pypit/traceslits.py]: These two are completely different.

```tc_dict``` is defined in the doc string (albeit not in great detail)

File: pypit/traceslits.py [hunk 16]
```diff
@@ -697,7 +698,7 @@ def _pca_pixel_slit_edges(self):
         # Step
         self.steps.append(inspect.stack()[0][3])
 
-    def remove_slit(self, rm_slits):
+    def remove_slit(self, rm_slits, TOL = 3.):
```
Comment (profxj) [inline pypit/traceslits.py]: Add to doc string

File: pypit/wavetilts.py [hunk 17]
```diff
@@ -255,11 +255,19 @@ def _trace_tilts(self, slit, wv_calib=None):
 
         """
         # Determine the tilts for this slit
+        tracethresh_in = self.settings['tilts']['tracethresh']
+        if isinstance(tracethresh_in,float) | isinstance(tracethresh_in,int):
```
Comment (profxj) [inline pypit/wavetilts.py]: use the ```isinstance``` syntax as above

File: pypit/core/arextract.py (hunk 2 above)
Comment (jhennawi) [reply pypit/core/arextract.py]: Done.

File: pypit/core/artraceslits.py (hunk 8 above)
Comment (jhennawi) [reply pypit/core/artraceslits.py]: Fixed.

File: pypit/core/artraceslits.py (hunk 9 above)
Comment (jhennawi) [reply pypit/core/artraceslits.py]: Read the docs! It is already there.

File: pypit/core/arskysub.py (hunk 6 above)
Comment (jhennawi) [reply pypit/core/arskysub.py]: My bad.

File: pypit/core/artraceslits.py (hunk 10 above)
Comment (jhennawi) [reply pypit/core/artraceslits.py]: That was a bug.

File: pypit/core/artraceslits.py (hunk 11 above)
Comment (jhennawi) [reply pypit/core/artraceslits.py]: I prefer native python for utility codes. This is a utility code.

File: pypit/traceslits.py (hunk 16 above)
Comment (jhennawi) [reply pypit/traceslits.py]: Done.

File: pypit/wavetilts.py (hunk 17 above)
Comment (jhennawi) [reply pypit/wavetilts.py]: Done.

**Comment by profxj** [comment]
Closing as per @jhennawi 's desire.

**Pull request closed without merging**